### PR TITLE
[Feature] Introducing OpenTelemetry to improve the observability of Doris.

### PR DESCRIPTION
**Pull request by luozenglin**
The collection of query traces is implemented in fe and be, and the spans are exported to zipkin.

# Proposed changes

Issue Number: #9577 

I plan to introduce OpenTelemetry to collect telemetry data, such as traces, metrics, logs, to improve the observability of Doris.

### The first phase uses OpenTelemetry to implement the collection and export of traces.

A trace represents the execution process of a single request in the system, span represents a logical operation unit with start time and execution duration in the system, and multiple spans form a trace.

1. support the collection and export of query traces

2. support the collection and export of load traces

## Problem Summary:

This pr implements the collection and sending of query traces. fe turns on traces, collects spans and propagates the traces information to be via prc. eventually each node exports the collected spans to zipkin. 

```
┌──────────────────────┐
│                   fe │
│                      │
│        span root     │
│            │         │
│    ┌───────┴────┐    ├────────────────┐
│    │            │    │                │
│  span A      span B  │                │
│                │     │                │
│                │     │                │
└────────────────┼─────┘                │
                 │                      │
           ┌─────┴───────────────┐      │
           │                     │      │
┌──────────┼───────────┐                │
│          │       be  │        be      │          ┌────────────┐
│          │           │                │          │            │
│         span C       │                │   span   │            │
│          │           ├────────────────┴──────────►   zipkin   │
│    ┌─────┴───────┐   │                           │            │
│    │             │   │                           │            │
│   span D      span E │                           │            │
│                      │                           └────────────┘
└──────────────────────┘
```

### Example：

1. Start zipkin:

```
curl -sSL https://zipkin.io/quickstart.sh | bash -s
java -jar zipkin.jar
```

2. Add Configuration：

fe.conf:
```
enable_tracing=true
trace_export_url=http://127.0.0.1:9411/api/v2/spans
```
be.conf:
```
enable_tracing=true
trace_export_url=http://127.0.0.1:9411/api/v2/spans
```

3.  Execute a query

4.  From zipkin UI：

截屏2022-05-17 12 20 32

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No)
3. Has document been added or modified: (No)
4. Does it need to update dependencies: (Yes)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by spaces-X** [comment]
Nice work!
i  just wonder what percentage of performance drawback will be caused by increasing the trace in general, looking forward to do a performance test

File: thirdparty/patches/opentelemetry-cpp-1.2.0.patch [hunk 1]
```diff
@@ -0,0 +1,374 @@
+diff --git a/api/include/opentelemetry/common/threadlocal.h b/api/include/opentelemetry/common/threadlocal.h
```
Comment (jackwener) [inline thirdparty/patches/opentelemetry-cpp-1.2.0.patch:1]: 😂 you git push the `patch` file.

**Comment by luozenglin** [comment]
> Nice work! i just wonder what percentage of performance drawback will be caused by increasing the trace in general, looking forward to do a performance test

Yeah, I will do performance testing when this feature is about to be completed. In order to reduce performance loss, performance optimization may be required at that time.

File: thirdparty/patches/opentelemetry-cpp-1.2.0.patch (hunk 1 above)
Comment (luozenglin) [reply thirdparty/patches/opentelemetry-cpp-1.2.0.patch:1]: This patch is to solve the problem that the opentelemtry-cpp library depends on `GLIBC_2.18`, and replaces the `thread_local` keyword in the opentelemtry-cpp library.

**Comment by SaintBacchus** [comment]
Will this feature collect the plan frament execution metrics as the sql profile?

**Comment by morningman** [comment]
> Will this feature collect the plan frament execution metrics as the sql profile?

Work in progress

**Pull request closed without merging**